feat(dockerfile): Add pip caching for faster build

V0lantis opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->
## Pip caching
For our corporation, we are doing a custom build. I noticed that every time, docker was downloading again and again the python packages and I thought it was a little bit too slow. I might want to enlarge this usage of pip caching for our CI as well.
 
wdyt?
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
feat(dockerfile): Add pip caching for faster build

potiuk reviewed (changes requested):
I would love to learn more about your use case and see how and whether it can help at all.

I am not sure if it changes anything and it adds some assumptions that the image will be built several times on the same machine while also for some reason it will not use the already cached layers.

Could you please show some usage where it actually helps and improve things ? Some benchmarks showing what you've done and before/after would be good.

Quite for sure caching when installing `pip` (first part of your change) has no effect. 

This is something that is only done once to upgrade to latest version and then (regardless of caching) `pip` does not get re-installed again, because it is already in the target version. And (unless you have some other scenario in mind) re-using cache will only actually work when you use the same docker engine. And in this case - unless you change your docker file parameters, it is already handled by docker layer caching. I.e. if you are building your docker image twice in the same docker engine with the same parameters, then caching is done at the "container layer" level. So when you run your build again, on the same docker engine you should not see it trying to install Python again. You should instead see that CACHED layer is used. And this is all without having to mount cache volume.

So I wonder why you would like to cache pip upgrade with cache.

The second case might be a bit more interesting - but there also you will see reinstallation only happening if you change requirements.txt. And yes - in this case it will be a bit faster when you just add one requirement, but this will also work for the same user when rebuilding the image over and over again on the same machine, which is generally not something that you should use airflow's dockerfile for (usually) - usually this can be done by extending the image (`FROM apache/airflow`)  and adding lines in your own Dockerfile (where you could use indeed --mount-type cache etc.).

I wonder if you could describe your case in more detail and show some benchmark example of how much time is saved for particular scenarios.

I am not totally against it, but I would love to understand your case, because a) maybe you do not gain as much as you think or b) maybe you are doing somethign that prevents you from using docker layer caching or c) maybe you are using the Dockerfile of Airlfow in unintended way.

Looking forward for more detailed explanation of your case.





So caching `pip`  install makes no sense at all.

Commit:
chore: Remove cache mounting for pip version install

Remove cache mounting during pip version install
because cache isn't necessary here

Commit:
feat: Create id based on the underlying image and arch

Commit:
feat(dockerfile): Add 2 new args for cache id

- Add AIRFLOW_PIP_VERSION for having a cache per pip version
- Add PIP_CACHE_EPOCH to be able to invalidate cache easily

Commit:
doc(docker-stack): Add doc explaining pip caching

potiuk reviewed (approved):
Fantastic.  Looks great . Thanks for being responsive. For changelog - I think we just need new line here: https://github.com/apache/airflow/blob/main/docs/docker-stack/changelog.rst#airflow-27

Add it for 2.7.3 please so that this change is cherry-pickable and we will have the last chance to merge it (2.7.3 is just being cut). 

UPDATE: I will add it to do it faster.

potiuk committed:
fixup! doc(docker-stack): Add doc explaining pip caching

hussein-awala reviewed (approved):
Great work!

Commit:
chore: apply and fix static check

Commit:
doc: Fix static checks

Commit:
chore(dockerfile): Move cache to tmp dir

Commit:
doc: move pip caching to 2.7.4

Commit:
feat(pip-cache): Pass pip cache path to env var
